File: src/main/java/edu/kit/ifv/mobitopp/actitopp/WRDDefaultModelStep.java
package edu.kit.ifv.mobitopp.actitopp;

/**
 * 
 * @author Tim Hilgert
 *
 * object for a wrd (weighted random draw) model step 
 */
public class WRDDefaultModelStep extends AbsHModelStep
{
	// category to get a random draw
	private String category;
	// activitytype for personal, activity type specific distributions
	private ActivityType activityType;
	
	// Distribution element that is used to pick a random number
	private WRDDiscreteDistribution wrddist;
	
	// decision if the distribution should be adapted after drawing 
	private boolean modifydistribution = false;
			
	// element that is finally chosen based on weighted random draw
  private int chosenDistributionElement;
       
  //limits the range in which a number should be picked randomly.
  private int lowerBoundLimiter = -1;
  private int upperBoundLimiter = -1;

  /**
   *
   * @param id
   * @param category
   * @param activityType
   * @param modelCoordinator
   */
  public WRDDefaultModelStep(String id, String category, ActivityType activityType, Coordinator modelCoordinator)
  {
    super(id, modelCoordinator);
    
    this.category = category;
    this.activityType = activityType;
    
    // check if a personalized distribution for this id, category and activity type already exists. If not, create one.
    this.wrddist = modelCoordinator.getpersonalWRDdistribution(id, category, activityType);
    if (wrddist==null) 
    {
    	WRDModelSteplnformation modelstep = modelCoordinator.getFileBase().getModelInformationforWRDStep(id);	
  		WRDModelDistributionInformation distributioninformation = modelstep.getWRDDistribution(category);
    	wrddist = new WRDDiscreteDistribution(distributioninformation);
    	
    	modelCoordinator.addpersonalWRDdistribution(id, category, activityType, wrddist);
    }
  }
  
  /**
   * creates wrd model step element without a given activity type
   * may be used when wrddist should not be dependent from activity type
   *
   * @param id
   * @param category
   * @param modelCoordinator
   */
  public WRDDefaultModelStep(String id, String category, Coordinator modelCoordinator)
  {
    this(id, category, ActivityType.UNKNOWN, modelCoordinator);
  }
  
  
  @Override
  public int doStep()
  {
  
    // pick a random number within the given boundaries
    chosenDistributionElement = wrddist.getRandomPickFromDistribution(this.lowerBoundLimiter, this.upperBoundLimiter, modelCoordinator.getRandomGenerator());
        
    if(modifydistribution)
    {
    	wrddist.modifydistributionelement(chosenDistributionElement);
    }     
    
    return getchosenDistributionElement();
  }
  

  public void printDecisionProcess()
  {
  	System.out.println("--------------- MC-Simulation @ " + this.id + this.category + this.activityType +" ---------------");
    System.out.println("From " + this.lowerBoundLimiter);
    System.out.println("To " + this.upperBoundLimiter);
    System.out.println("Random Value: " + modelCoordinator.getRandomGenerator().getLastRandomValue());
    System.out.println("Chosen: " + chosenDistributionElement);
    System.out.println("");
  }
   
  /**
   * 
   * @param lowerbound
   * @param upperbound
   */
  public void setRangeBounds(int lowerbound, int upperbound)
  {
    this.lowerBoundLimiter = lowerbound;
    this.upperBoundLimiter = upperbound;
  }

  public int getchosenDistributionElement()
  {
    return chosenDistributionElement;
  }  
  
	public void setModifydistribution(boolean modifydistribution) 
	{
		this.modifydistribution = modifydistribution;
	}

}

File: src/main/java/edu/kit/ifv/mobitopp/actitopp/WRDDefaultModelStep.kt
package edu.kit.ifv.mobitopp.actitopp

import edu.kit.ifv.mobitopp.actitopp.changes.Category
import edu.kit.ifv.mobitopp.actitopp.enums.ActivityType

/**
 * @author Tim Hilgert
 *
 *
 * object for a wrd (weighted random draw) model step
 */
class WRDDefaultModelStep(
    id: String, // category to get a random draw
    private val category: Category, // activitytype for personal, activity type specific distributions
    private val activityType: ActivityType = ActivityType.UNKNOWN, // surrounding modelCoordinator for this step
    private val modelCoordinator: Coordinator
) :
    AbsHModelStep(id) {

    // Distribution element that is used to pick a random number

    // check if a personalized distribution for this id, category and activity type already exists. If not, create one.
    private val weightedDistribution: WRDDiscreteDistribution =
        modelCoordinator.getOrCreateWeightedDistribution(id, category, activityType)

    // decision if the distribution should be adapted after drawing
    private var modifydistribution = false


    private var bounds: IntRange = Int.MIN_VALUE..Int.MAX_VALUE



    public override fun doStep(): Int {
        // pick a random number within the given boundaries

        val selection = weightedDistribution.getRandomPickFromDistribution(
            this.bounds, modelCoordinator.randomGenerator
        )

        if (modifydistribution) {
            weightedDistribution.modifydistributionelement(selection)
            weightedDistribution.theModifiee = modelCoordinator.person
        }

        return selection.also { chosenDistributionElement = selection }
    }

    @Deprecated("Only used in the print method below")
    private var chosenDistributionElement = 0

    fun printDecisionProcess() {
        println("--------------- MC-Simulation @ " + this.id + this.category + this.activityType + " ---------------")
        println("From " + this.bounds.first)
        println("To " + this.bounds.last)
        println("Random Value: " + modelCoordinator.randomGenerator.lastRandomValue)
        println("Chosen: $chosenDistributionElement")
        println("")
    }

    /**
     * @param lowerbound
     * @param upperbound
     */
    fun setRangeBounds(lowerbound: Int, upperbound: Int) {
        require(lowerbound <= upperbound) {
            "Cannot set a range to $lowerbound $upperbound somehow the upperbound is smaller than the lower bound"
        }
        this.bounds = lowerbound..upperbound
    }

    fun setModifydistribution(modifydistribution: Boolean) {
        this.modifydistribution = modifydistribution
    }
}
